packages/playwright-core/src/server/injected/webSocketMock.ts

```diff
@@ -0,0 +1,345 @@
+        this.removeEventListener('error', this._onerrorListener);
+      this._onerrorListener = listener;
+      if (this._onerrorListener)
+        this.addEventListener('error', this._onerrorListener);
+    }
+
+    get onopen() {
+      return this._onopenListener;
+    }
+
+    set onopen(listener) {
+      if (this._onopenListener)
+        this.removeEventListener('open', this._onopenListener);
+      this._onopenListener = listener;
+      if (this._onopenListener)
+        this.addEventListener('open', this._onopenListener);
+    }
+
+    get onmessage() {
+      return this._onmessageListener;
+    }
+
+    set onmessage(listener) {
+      if (this._onmessageListener)
+        this.removeEventListener('message', this._onmessageListener as any);
+      this._onmessageListener = listener;
+      if (this._onmessageListener)
+        this.addEventListener('message', this._onmessageListener as any);
+    }
+
+    send(message: WebSocketMessage): void {
+      if (this.readyState === WebSocketMock.CONNECTING)
+        throw new DOMException(`Failed to execute 'send' on 'WebSocket': Still in CONNECTING state.`);
+      if (this.readyState !== WebSocketMock.OPEN)
+        throw new DOMException(`WebSocket is already in CLOSING or CLOSED state.`);
+      if (this._passthrough)
+        this._apiSendToServer(message);
+      else
+        messageToData(message, data => binding({ type: 'onMessageFromPage', id: this._id, data }));
+    }
+
+    close(code?: number, reason?: string): void {
+      if (code !== undefined && code !== 1000 && (code < 3000 || code > 4999))
+        throw new DOMException(`Failed to execute 'close' on 'WebSocket': The close code must be either 1000, or between 3000 and 4999. ${code} is neither.`);
+      if (this.readyState === WebSocketMock.OPEN || this.readyState === WebSocketMock.CONNECTING)
+        this.readyState = WebSocketMock.CLOSING;
+      if (this._ws)
+        this._ws.close(code, reason);
+      else
+        this._onWSClose(code, reason, true);
+    }
+
+    // --- methods called from the routing API ---
+
+    _apiEnsureOpened() {
+      // This is called at the end of the route handler. If we did not connect to the server,
+      // assume that websocket will be fully mocked. In this case, pretend that server
+      // connection is established right away.
+      if (!this._ws)
+        this._ensureOpened();
+    }
+
+    _apiSendToPage(message: WebSocketMessage) {
+      // Calling "sendToPage()" from the route handler. Allow this for easier testing.
+      this._ensureOpened();
+      if (this.readyState !== WebSocketMock.OPEN)
+        throw new DOMException(`WebSocket is already in CLOSING or CLOSED state.`);
+      this.dispatchEvent(new MessageEvent('message', { data: message, origin: this._origin, cancelable: true }));
+    }
+
+    _apiSendToServer(message: WebSocketMessage) {
+      if (this.readyState === WebSocketMock.CLOSING || this.readyState === WebSocketMock.CLOSED)
```

Comment:
Can this be removed and instead the check at line 259 be relaxed, so that we get error from _ws?